bookstore/tests/test_handlers.py

```diff
@@ -1,6 +1,87 @@
+
 import pytest
 
+from tornado.testing import AsyncTestCase
+from tornado.web import Application, HTTPError
+from tornado.httpserver import HTTPRequest
+from traitlets.config import Config
+
+from ..bookstore_config import BookstoreSettings
+from ..handlers import BookstoreVersionHandler, build_settings_dict
+
+
+@pytest.fixture(scope="class")
+def bookstore_settings(request):
+    mock_settings = {
+        "BookstoreSettings": {"s3_access_key_id": "mock_id", "s3_secret_access_key": "mock_access"}
+    }
+    config = Config(mock_settings)
+    bookstore_settings = BookstoreSettings(config=config)
+    if request.cls is not None:
+        request.cls.bookstore_settings = bookstore_settings
+    return bookstore_settings
+
+
+def test_build_settings_dict(bookstore_settings):
+    expected = {
+        'validation': {'archive_valid': True, 'bookstore_valid': False, 'publish_valid': True},
+        'version': '2.3.0.dev',
+    }
+    assert expected == build_settings_dict(bookstore_settings)
+
+
+@pytest.mark.usefixtures("bookstore_settings")
+class TestCloneAPIHandler(AsyncTestCase):
+    def setUp(self):
+        super().setUp()
+
+        self.mock_application = Mock(
+            spec=Application,
+            ui_methods={},
+            ui_modules={},
+            settings={"bookstore": build_settings_dict(self.bookstore_settings)},
+            transforms=[],
+        )
+
+    def get_handler(self, uri, app=None):
+        if app is None:
+            app = self.mock_application
+        connection = Mock(context=Mock(protocol="https"))
+        payload_request = HTTPRequest(
+            method='GET',
+            uri=uri,
+            headers={"Host": "localhost:8888"},
+            body=None,
+            connection=connection,
+        )
+        return BookstoreVersionHandler(app, payload_request)
+
+    def test_get(self):
+        """This is a simple test of the get API at /api/bookstore
+
+        The most notable feature is the need to set _transforms on the handler.
+
+        The default value of handler()._transforms is `None`.
+        This is iterated over when handler().flush() is called, raising a TypeError.
+        
+        In normal usage, the application assigns this when it creates a handler delegate.
+
+        Because our mock application does not do this 
+        As a result this raises an error when self.finish() (and therefore self.flush()) is called.
+
+        At runtime on a live Jupyter server, application.transforms == [].
+        """
+        get_handler = self.get_handler('/api/bookstore/')
+        setattr(get_handler, '_transforms', [])
+        get_handler.get()
 
-def test_handlers():
-    pass
+    def test_build_response(self):
+        empty_handler = self.get_handler('/api/bookstore/')
+        expected = {
+            'bookstore': True,
+            'validation': {'archive_valid': True, 'bookstore_valid': False, 'publish_valid': True},
+            'version': '2.3.0.dev',
```

Comment:
Will this test have to be updated each time the version is bumped? Might want to import the version for building this test object

Reply:
That's an excellent suggestion!